### ISSUE TITLE
Question: usage with parallel request on a single database instance

### ISSUE THREAD

**Issue by seanjyjy** [romeerez/pg-transactional-tests#2]
Hi @romeerez,

Thank you for creating this library—it's been incredibly helpful!

I have a question regarding test scenarios with multiple requests (two or more) and transaction handling. In my current use case, I’m testing for race conditions by sending two parallel requests. However, I’ve run into an issue where both requests share the same transaction. When one request fails, it throws the expected error, but because both requests are tied to the same transaction, the other request will ultimately becomes a transaction rollback error instead of a success case.

You can view the current scenario as
```js
const request1 = supertest(app).call() // first http request
const request2 = supertest(app).call() // second http request
const [response1, response2] = await Promise.all([request1, request2])

// sample expectation, can be vice versa. 
expect(response1.statusCode).toBe(200) // However instead of getting 200, it will get the error of transaction rollback
expect(response2.statusCode).toBe(4xx)
```

Is there a way to configure the library so that each request can have its own transaction within the same test? If that’s not possible, could you offer any advice on how to test parallel requests and ensure that the original errors are captured, rather than the transaction rollback error?

I appreciate any insights you can provide!

**Comment by romeerez** [comment]
I'm very glad for the library to find its happy users!

I think it's possible to support with a help of [AsyncLocalStorage](https://nodejs.org/api/async_context.html#new-asynclocalstorage):

```ts
import { parallelTransaction } from 'this-library'

const request1 = parallelTransaction(() => supertest(app).call())
const request2 = parallelTransaction(() => supertest(app).call())
```

`parallelTransaction` will do `AsyncLocalStorage.run` under the hood to define a new context, and in the moment of query it will know that one query belongs to transaction 1, and the other query to transaction 2, and they both should have different connections.

Seems to be pretty doable.

But, are you sure you want this functionality? It's a bit unusual to test race conditions, because they're unpredictable, maybe 1st will finish earlier, and maybe 2nd, and results may change between test runs . I've never seen such tests.

**Comment by seanjyjy** [comment]
> parallelTransaction will do AsyncLocalStorage.run under the hood to define a new context, and in the moment of query it will know that one query belongs to transaction 1, and the other query to transaction 2, and they both should have different connections.

Do let me know if my thought process is in the right track, I suspect the issue stems from both request sharing the same pg monkey patch under the hood, hence both share the same state as currently for the same DB, i presume they will be using the same client. As a result, they both initiate the request under the same transaction (aka `begin`), and if one of the query fails due to a constraint violation, the entire transaction is rolled back. I've attempted to resolve this using AsyncLocalStorage, but without success. I might be trying it wrongly and if so, could you kindly point me to the right direction? If you'd like, I can prepare a small reproducible example to further investigate the issue.

> But, are you sure you want this functionality? It's a bit unusual to test race conditions, because they're unpredictable, maybe 1st will finish earlier, and maybe 2nd, and results may change between test runs . I've never seen such tests.

The purpose of this test is to ensure that our constraints are properly checked and enforced, serving as a sanity check. You're correct that the order of completion for the first and second queries might vary, which is why this Jest test essentially asserts that one query succeeds while the other fails. (it checks for both combinations)

Again thanks for the swift replies @romeerez

**Comment by romeerez** [comment]
Yes, you're correct, it's designed to always utilize a single connection, and does monkey-patching, because a single transaction can only operate on a single connection, and the lib wasn't designed for parallel transactions.

> I've attempted to resolve this using AsyncLocalStorage, but without success.

Sure, the changes had to be made in the library, because it patches `pg`, and it's not trivial to control it from outside.

I published update in v1.2.0:
- changed exports, so now you should import `testTransaction` and it has methods `start`, `rollback`, `close`. I decided it's a good idea, so in your code it's obvious this is a transaction meant for tests, not to be used in the app.
- added `testTransaction.parallel` to support your case.

[Here is a section](https://github.com/romeerez/pg-transactional-tests?tab=readme-ov-file#parallel-transactions) with example.

In the readme you can see that usage is slightly simplified. Now it patches automatically when calling `start`, and unpatches automatically when all transactions were rolled back, or when calling `close`.
